File: labs/mobile/donor-archive/tunnel/SubnetRangeScout.kt
package com.luminet.android.tunnel

data class SubnetScoutTarget(
    val ip: String,
    val rttMs: Long,
    val isClean: Boolean
)

data class SubnetScoutReport(
    val subnet: String,
    val totalTested: Int,
    val cleanCount: Int,
    val bestTargets: List<SubnetScoutTarget>
)

class SubnetRangeScout(
    val maxRttMsThreshold: Long = 200L
) {
    fun scout(baseSubnet: String, count: Int = 10): SubnetScoutReport {
        val targets = mutableListOf<SubnetScoutTarget>()
        var clean = 0
        for (i in 1..count.coerceAtMost(254)) {
            val ip = "$baseSubnet.$i"
            val rtt = (20L + (i * 11L) % 150L)
            val isClean = rtt <= maxRttMsThreshold
            if (isClean) clean++
            targets.add(SubnetScoutTarget(ip, rtt, isClean))
        }
        val best = targets.sortedBy { it.rttMs }.take(5)
        return SubnetScoutReport(
            subnet = "$baseSubnet.0/24",
            totalTested = targets.size,
            cleanCount = clean,
            bestTargets = best
        )
    }
}

File: labs/control-ui/api-shadow/subnet_range_scout.ts
export interface ScoutedTarget {
  ip: string;
  rttMs: number;
  isClean: boolean;
}

export interface SubnetScoutReport {
  subnet: string;
  testedCount: number;
  cleanCount: number;
  bestTargets: ScoutedTarget[];
}

export function scoutSubnetRange(baseSubnet: string, count: number = 8, thresholdMs: number = 200): SubnetScoutReport {
  const targets: ScoutedTarget[] = [];
  let clean = 0;

  for (let i = 1; i <= Math.min(count, 254); i++) {
    const ip = `${baseSubnet}.${i}`;
    const rttMs = 15 + ((i * 13) % 180);
    const isClean = rttMs <= thresholdMs;
    if (isClean) clean++;
    targets.push({ ip, rttMs, isClean });
  }

  targets.sort((a, b) => a.rttMs - b.rttMs);

  return {
    subnet: `${baseSubnet}.0/24`,
    testedCount: targets.length,
    cleanCount: clean,
    bestTargets: targets.slice(0, 5),
  };
}
